### ISSUE TITLE
How is grad norm computed

### ISSUE THREAD

**Issue by LukeBailey181** [thinking-machines-lab/tinker-cookbook#331]
Hi! Im training gpt-oss-120b with CISPO loss. I get the following grad norm returned on a regular batch of on policy data:

`OptimStepResponse(metrics={'unclipped_grad_l2:mean': 13468.94921875})`

How exactly is `unclipped_grad_l2:mean` computed, as the reported grad norm is dramatically too large.

Thanks in advance for any help with this!

**Comment by Irvingwangjr** [comment]
I think tinker use loss:sum, so if you want to do loss-mean, you need to do that using weights in loss_fn

**Comment by YujiaBao** [comment]
Hey! Good question — the short answer is that the grad norm is large because all Tinker losses (including CISPO) use sum reduction over tokens, not mean. So the gradient scales with the number of tokens in your batch, and for a model like gpt-oss-120b with long sequences, grad norms in the thousands aren't surprising.

As for `unclipped_grad_l2:mean` — the `:mean` part refers to how the metric is *aggregated* across microbatches and replicas (it gets averaged), not to how the loss is reduced. The underlying loss is still a sum.

As @Irvingwangjr pointed out, if you want mean-like behavior you'll need to account for it yourself. The way to do this with CISPO is through the `advantages` tensor you pass in `loss_fn_inputs`. Since the CISPO loss computes:

```python
loss = -(clipped_ratio.detach() * target_logprobs * advantages).sum()
```

you can normalize your advantages by the number of action tokens per datum. For example, in the cookbook's RL data processing ([`tinker_cookbook/rl/data_processing.py`](https://github.com/thinking-machines-lab/tinker-cookbook/blob/main/tinker_cookbook/rl/data_processing.py#L163)), advantages are broadcast uniformly across action tokens:

```python
advantages = [0] * ob_len + [traj_advantage] * len(action_tokens)
```

To get mean-like scaling, you could divide by the number of action tokens:

```python
per_token_advantage = traj_advantage / len(action_tokens)
advantages = [0] * ob_len + [per_token_advantage] * len(action_tokens)
```

This is also called out in the [losses docs](https://github.com/thinking-machines-lab/tinker-cookbook/blob/main/docs/losses.mdx): "for all objectives we sum the token-level losses over the sequence length... If you would like to explore different aggregation schemes, you can include that in the advantage tensor computation."
